### ISSUE TITLE
Torlock - andmatch filter on Sonarr 3 but not on Manual Search

### ISSUE THREAD

**Issue by rambius1** [Jackett/Jackett#8215]
**OS**: Windows 10 1809

**Jackett Version**: 0.16.18.0

**Last Working Jackett Version**: N/A

**Are you using a proxy or VPN?** Yes, NordVPN.

### Description

Sonarr - 3.0.3.754

andmatch filtered out results, most likely due to characters length of title name.  

Can the characters length be adjusted for some series with long title name?

Same issue from Torlock and 1337x. 

Only affect searches in Sonarr, but error is with Jackett (see debug logs).

Jackett's manual search for "Zoey's Extraordinary Playlist" works fine and list "1080p AMZN WEB-DL x265" version in the result.

Sonarr's manual search for "Zoey's Extraordinary Playlist" via Torznab indexer does not work and no ""1080p AMZN WEB-DL x265"  version in the result.

Jackett's debug log show "andmatch filter" is the cause of the issue. 

### Logged Error Messages

CardigannIndexer (torlock): skipping Zoey's Extraordinary Playlist (2020) S01E01 Pilot (1080p AMZN WEB-DL x265 10bit EAC3 5.1 - HxD) [TAoE] (andmatch filter)

CardigannIndexer (torlock): skipping Zoey's Extraordinary Playlist (2020) S01E03 Zoey's Extraordinary Boss (1080p AMZN WEB-DL x265 10bit EAC3 5.1 - HxD) [TAoE] (andmatch filter)

**Comment by garfield69** [comment]
The issue here is that Sonarr3 is making a request for `Zoeys Extraordinary Playlist S01E01` to Jackett, and the Tolock Indexer is rejecting `Zoey's Extraordinary Playlist S01E01` because it does not match.
The andmatch operates on all requests whether then come from the Jackett Dashboard manual search, or via requests via the Torznab API.
We put the Torlock indexer andmatch in place because the Torlock web site search engine operates on multiple keywords with a OR instead of an AND and thus returns spurious results.
You cannot have it both ways, its other lots off useless results (which if ther eare enough of them returned will force to good ones off the page returned and you miss out of hits)
or we andmatch so we minimise miss-hits but it does mean titles with apostrophe's will be missed.

use another indexer.

**Comment by garfield69** [comment]
@cadatoiva  do you think we can get Cardigann to strip the apostrophe(s) from both sides of the word string comparison so that the andmatch will effectively match `Zoe's` and `Zoes` for instance?

**Comment by cadatoiva** [comment]
I think this should be for our internal and matching in general, not just Cardigann. It would be doable, but should also be more than just apostrophe(s) I think. We could use [`char.IsLetterOrDigit()`](https://docs.microsoft.com/en-us/dotnet/api/system.char.isletterordigit?view=netframework-4.8) to strip everything that isn't a letter or digit, regardless of language. Or maybe use [`char.IsPunctuation()`](https://docs.microsoft.com/en-us/dotnet/api/system.char.ispunctuation?view=netframework-4.8) for just punctuation?

The screenshot below is the output of every unicode value that matches `char.IsPunctuation()` according to MS documentation (sorry about the boxes, my computer doesn't unicode well)
![image](https://user-images.githubusercontent.com/2086652/79406230-1ee0f700-7f5c-11ea-8b0a-ebf087ce303f.png)

And this is a screenshot of (some) of the output of `char.IsLetterOrDigit()` to give you an idea of what it would match 
![image](https://user-images.githubusercontent.com/2086652/79406415-a464a700-7f5c-11ea-8d01-a22c83b8e9d1.png)

**Comment by garfield69** [comment]
um, hard-coding may not be a good idea with such a large set.
Not all indexers are English. We need to allow  Greek, Russian and Simplified Chinese to give some examples of foreign indexers.
But one possibility that occurs to me, may be if the andmatch took an optional args statement  where we can list what we want to ignore

So default with no args is a straight one to one match as usual
```yaml
    rows:
      selector: div.hitem
      filters:
        - name: andmatch
```
and args provides single characters or groupsets that should be ignored (stripped) before the compare takes place
such as for example:
```yaml
    rows:
      selector: div.hitem
      filters:
        - name: andmatch
          args: "$#&()'"
```
```yaml
    rows:
      selector: div.hitem
      filters:
        - name: andmatch
          args: "\\p{IsCJKUnifiedIdeographs}"
```
or something similar.

**Comment by cadatoiva** [comment]
The two functions I mentioned above base themselves off the unicode standard, so it will strip all unicode punctuation from all language sets, or keep all letters from all language sets, or however we decide. We don't hard code it specifically, there's a set of microsoft provided functions that we can use to say "is it punctuation in any language? skip it" or similar. `char.IsLetterOrDigit()` and `char.IsPunctuation()` are the C# native features similar to `"\\p{IsCJKUnifiedIdeographs}"` 

Specifically `char.IsLetterOrDigit()` is equal to `\w` https://docs.microsoft.com/en-us/dotnet/standard/base-types/character-classes-in-regular-expressions#word-character-w

And `char.IsPunctuation()` is equal to `"\\p{P}"`

I think global stripping of characters is likely going to be a better choice than having to specify for each tracker.

**Comment by garfield69** [comment]
I see. sounds reasonable. 
then perhaps the args can be just a switch so that we can choose strict matching vs ignore punctuation?
because I guarantee if we go the by-default-strip-punctuation we will come across a web site where that will cause an issue.

**Comment by cadatoiva** [comment]
Yeah, we can add something like an optional arg "strict" but it defaults to stripping:
```YAML
filters:
  # doesn't strip punctuation
  - name: andmatch 
    args: strict
  # strips punctuation
  - name: andmatch
```

And where this change would have to be made also means that the C# trackers will get this same upgrade.

**Comment by ngosang** [comment]
Related #7673

**Comment by rambius1** [comment]
Will the suggested fix (mentioned above) add to upcoming build?

**Comment by rambius1** [comment]
OK, but this apostrophe issue is not just with Torlock indexer.  The same torrent files exist on 1337x and ETTV. I loaded and tested the other 2 indexers in Jackett with same failed result.

**Comment by cadatoiva** [comment]
> OK, but this apostrophe issue is not just with Torlock indexer. The same torrent files exist on 1337x and ETTV. I loaded and tested the other 2 indexers in Jackett with same failed result.

The way I'll be implementing this fix will default to being used on every indexer and yml definition, to be turned off if it causes worse problems on a case by case basis. It's on my list, so we'll get this fixed.

**Comment by ilike2burnthing** [comment]
Proposed fix - https://github.com/Jackett/Jackett/issues/8246

See also - https://github.com/Sonarr/Sonarr/pull/3684

**Comment by Taloth** [comment]
@ilike2burnthing No it's not. The andmatch filter strictness is largely unrelated to Sonarr sending the full series titles. Indexers can have small variations and titles on tvdb can change (for example include a year) and thus the andmatch filter should be able to handle that.
Sonarr providing the full title certainly alleviates some of these cases, but it doesn't fix the underlying problem.

**Comment by ilike2burnthing** [comment]
Did I misunderstand that in the case of "Zoey's Extraordinary Playlist", Sonarr is currently stripping the apostrophe, but were `title` to be used with your PR, Sonarr would no longer do that, thus returning results?

**Comment by Taloth** [comment]
It just hides the problem, any result without the apostrophe would be filtered out. Same problem, just in reverse. It doesn't fix the underlying issue. (it only helps for indexers that don't have a fuzzy search function or only a series title lookup).
That's why the strictness of the andmatch is the real challenge. A filter using levenshtein distance is probably a better pick here since it's intended to do what you want: filter out stuff that's way too different compared to the search query.
Removing special characters in the andmatch will probably do the trick too.
